### ISSUE TITLE
Add declarative annotations for "issue" type support and infrastructure

### ISSUE THREAD

**Issue by mkobit** [junit-pioneer/junit-pioneer#135]
For example, Spock has [`@Issue`](http://spockframework.org/spock/javadoc/1.0/spock/lang/Issue.html) which is used to

> Indicates that a feature method or specification relates to one or more issues in an external issue tracking system.

Which might look like

```groovy
@Issue("https://github.com/junit-pioneer/junit-pioneer/issues/131")
```

It can often be useful to annotate test with some external issue for reporting, conditionally enabling/disabling, or to build additional tooling on.

Shorthand may also be desired if the project itself has some known destination:

```groovy
@Issue("junit-pioneer/junit-pioneer/issues/131")
```

It would be nice to have some kind of support for annotating code with issue references such that they can used for some of the previous use cases.

A few cases to consider for the API definition and consideration:

* Arbitrary issue URL - `@Issue("https://github.com/junit-pioneer/junit-pioneer/issues/131")`
* Arbitrary issue URL with prefix coming from some other configuration -  `@Issue("junit-pioneer/junit-pioneer/issues/131")`
* Public GitHub shorthand - `@GitHub("junit-pioneer/junit-pioneer", 131)`
* Private GitHub enterprise, JIRA, GitLab, etc.
* Meta-annotated issue - `@JUnitPioneerIssue(131)`

**Comment by nicolaiparlog** [comment]
If this is just about showing the URL for an issue, a comment would do just fine. I would only use an annotation if the issue could somehow be relevant for test execution, but I struggle to see a use case for that. Do you have any idea, @mkobit?

**Comment by marcphilipp** [comment]
Related issue: https://github.com/junit-team/junit5/issues/1900

**Comment by nicolaiparlog** [comment]
Thanks for pointing that out. So from what I gathered, there are two possible features:

* add given issue to report
* gather all test results for an issue and report them

The former is trivial and we could do that. The latter sounds much more interesting! I'm imagining a dynamically generated test tree where each issue is a node and each test belonging to that issue is a subnode. By storing the results of the original tests when they were run, the generated tests could trivially pass or throw the same exception, effectively giving a second view on the same test results. Thing is, we'd need a way to guarantee that this test tree is generated at the end of a run when all results are in - is there a way to do that?

Feature set aside, I see three different ways going forward:

1. do whatever we can now, pioneering (get it?) a feature that may become part of JUnit 5 at some point
2. wait for JUnit 5 to decide on a report format and then do whatever we can
3. wait for JUnit 5 to decide the fate of junit-team/junit5#1900 and do whatever we can that isn't yet covered

The second option doesn't seem to make much sense to me because I don't think Pioneer is impacted by JUnit's report format. Looks like it's going to be 1. or 3.

**Comment by marcphilipp** [comment]
I think "gather all test results for an issue and report them" should be done by a consuming reporting tool/generator. We'll just have to get that report format out...

**Comment by sormuras** [comment]
1.7...

**Comment by nicolaiparlog** [comment]
Ok, so unless users are clamoring for somehow gathering test results before the report format is out, let's table that discussion.

In the meantime, if anybody wants to add an annotations that reports a URL to the report, they can start working on it - I'll make this issue _up for grabs_. That said, the annotation's name still needs to be decided.

**Comment by Bukama** [comment]
As I opened junit#1900 I want to add some thoughts to this (Writing on phone so format will be most probably  bad - I‘ll fix this when back on desk tomorrow or the day after it):

I also see that a report builder and the annotation itselfs are two different things that can and should be implemented seperated from each other. 

As @marcphillipp mentions here / in junit5#1900 the report format ( https://github.com/ota4j-team/opentest4j/issues/9 ) has still to be definied. Until then, as far as I understand marcphillip, the JUnit-Team does not want to implement an Issue/Requirement-annotation but a generic annotation. Therefore they declined my PR which implemented a „@Requirement“ annotation. 

I can understand the wish of the team to implement generic annotations but as one who also worked as a QA I still see a huge need in „Requirement/ Issue“-Annotation as a core annotation, because requirements and (them following) avoidance of technical errors are the absoluty base of software developement.

This said I would really love to see such annotation be implemented. For the name I could go with „@Issue“ and /  or „@Requirement“. Why do I write and/or? Because I see a different focus on the words. A requirement is a needed *professional* functionality, while issues are in most cases connected with bugs (in general) or the avoidance of technical problems. I know many tools, like JIRA, use the term for all kinds of „things to be done“, but I still see the use / eligibility of different „types“ according to the view on a test. I could also imagine that a filtering in a report, e.g. when I want to know if the tests for all requirements are fullfilled. 


As annotations can‘t be subclassed maybe some kind of „type“ attribute could be usefull, but hell not absolutly necessary. 

What I don‘t see is the need of putting a full URL into in annotation. In my opinion this is somethibg a report viewer (not even the builder) should prefix. Sure that would make the need that all things are in the same tool, which they are not very often. But even here there could be an attribute in the annotation if theres a huge need for the URL. 

To come to an end I suggest to create an „@Issue“-annotation which always needs an alphanumeric „id“-attribute, making it: „@Issue(„Req-123“)“ for example. 

As written, addional attributes for type and URL could also be usefull, making it „@Issue(id=„Req-123“, Type=IssueTypeEnum.REQUIREMENT, url=„https://path.to/my/issuetracker/issuie.php?id=Req-123“) or something like this.

**Comment by nicolaiparlog** [comment]
Thanks for your thoughts, @Bukama. I understand your qualms about the annotation's name. I think `@Requirement` would be too... specific (?) and think `@Issue` is better. If you think that sounds too much like a bug, we could use `@Ticket` as well.

Regarding the attributes:

* I like that you split `id` and `url` - that's a good idea. Wouldn't use `id`, though, but `value`, so it can be used without naming it, i.e. `@Ticket("123")`.
* Having an (optional) `url` attribute there is good. But it would be nice if the user could configure a URL pattern like `https://path.to/my/issuetracker/issuie.php?id=Req-%ID`, so we could automatically generate a URL for the output. That would be easiest if extensions can read [configuration parameters](https://junit.org/junit5/docs/current/user-guide/#running-tests-config-params), but I don't remember right now, whether they can.
* The issue type is a little more complicated. Every project has their own issue structure and we can't possibly represent them all. If (!) we want the type here, it should be an interface `TicketType`. I could see two methods:
    * `String name()` returns something like `"Bug"`.
    * `String message(String url)` creates a message that becomes part of the output. This way users have free reign over what shows up in the messages (good for filtering et al). Hell, they could even query the URL to include tags and whatnot if they don't mind the test's run time.

  To make things easier for simple cases, we can create `TicketType` implementations (probably as nested classes of the interface itself) that covers things like bugs, features, ...

What do you think?

**Comment by sormuras** [comment]
> [...] we could automatically generate a URL for the output. That would be easiest if extensions can read configuration parameters, but I don't remember right now, whether they can.

Yes, they can: https://junit.org/junit5/docs/current/api/org/junit/jupiter/api/extension/ExtensionContext.html#getConfigurationParameter(java.lang.String)

**Comment by marcphilipp** [comment]
@Bukama FWIW You can implement this today with a custom extension that reads your annotation and publishes it as a report entry. Then, you can register a `TestExecutionListener` (e.g., via service loader) that receives the report entry and writes it to your custom report.

**Comment by Bukama** [comment]
If #182 gets merged you, @nicolaiparlog , could assign this issue to me. I'll then start another run on it with a little more experience now.

**Comment by Bukama** [comment]
As I opened a draft (#189) I'll assign this to myself

**Comment by Bukama** [comment]
> As I opened a draft (#189) I'll assign this to myself

I closed the draft as I noticed I'm far away from done and also blocked by a possible bug in the TEL. Created an issue at JUnit ([Junit 5 - #2277](https://github.com/junit-team/junit5/issues/2277)).
